app/models/authentication.rb

```diff
@@ -142,10 +136,16 @@ def assign_values(options)
     self.attributes = options
   end
 
-  def self.build_credential_options
-    CREDENTIAL_TYPES.each_with_object({}) do |(k, v), hash|
-      hash[k] = v.constantize.descendants.each_with_object({}) do |klass, fields|
-        fields[klass.name] = klass::API_OPTIONS if defined? klass::API_OPTIONS
+  def self.credential_types
+    credential_subclasses = descendants.select { |d| d.try(:credential_type) }.sort_by(&:name)
```

Comment:
Can you use `.method_defined?` here so we don't actually invoke the credential_types call
```suggestion
    credential_subclasses = descendants.select { |d| d.method_defined?(:credential_type) }.sort_by(&:name)
```

Reply:
`method_defined?` gives an empty response here, might be due `credential_type` being a class method
```
3.0.1 :002 > credential_subclasses = Authentication.descendants.select { |d| d.method_defined?(:credential_type) }.sort_by(&:name)
 => [] 
3.0.1 :003 > credential_subclasses = Authentication.descendants.select { |d| d.try(:credential_type) }.sort_by(&:name)
 => [ManageIQ::Providers::AnsibleTower::AutomationManager::AmazonCredential(id: integer, name: string, authtype: string, userid: string,
```

Reply:
Wow you're right that's very weird....
```
>> ManageIQ::Providers::ExternalAutomationManager::Authentication.credential_type
=> "external_credential_types"
>> ManageIQ::Providers::ExternalAutomationManager::Authentication.method_defined?(:credential_type)
=> false
```

Okay this is fine then but I have no idea why method_defined? isn't working